stdlib/public/core/StringGuts.swift

```diff
@@ -23,38 +23,92 @@ import SwiftShims
 @_fixed_layout
 public // FIXME
 struct _StringGuts {
-  // TODO 32-bit: Insert padding between fields
   public // FIXME for testing only
-  var _storage: (_StringObject, UInt)
+  var _object: _StringObject
 
+  public // FIXME for testing only
+  var _otherBits: UInt // (Mostly) count or inline storage
+
+#if arch(i386) || arch(arm)
+  public // FIXME for testing only
+  var _extraBits: UInt // Start address or inline storage
```

Comment:
To avoid littering `#if` all over `StringGuts`, would it make sense to make a `StringExtraBits` trivial wrapper struct here? Kind of how `StringObject` helped abstract over BridgeObject usage? This might also simplify some of `StringGuts` proper.

Reply:
Do you mean replacing `UInt` with an `_ExtraBits` struct that's empty on 64-bit platforms?  That could work! (Perhaps `_otherBits` could live there too.) I'll try it and see if it makes things prettier.

The general issue is that `_StringGuts` layouts on 32-bit and 64-bit platforms are divergent enough that I find it difficult to unify them in a satisfying way.

Reply:
My `_ExtraBits` experiment was unsuccessful -- I eliminated a couple of minor `#if`s, but the majority remained, and I felt the wins didn't justify the additional complexity of an extra struct. ;(

Currently, `_StringObject` isn't great on 32-bit platforms: we don't have tagged pointers, so even with a 64-bit representation, we only have 29 bits free for a payload (assuming we reserve 3 bits for flags). This isn't enough to store an unmanaged base pointer, so I had to reroute that to `_extraBits` -- adding a bunch of `#if`s in `_StringGuts`. (A native tagged object reference type would help here a lot.)

Reply:
Another idea: moving `_extraBits` to `_StringObject`, so that it has feature parity with the 64-bit variant. It seems weird to extend `_StringObject` to three words, but it would be an improvement overall.

Some `#if`s would (probably) remain in StringGuts.swift, though, since 32-bit `_StringObject` would have space to store the start address of all contiguous strings.

Reply:
I think, for now, making `_StringObject` 12-byte on 32-bit would make sense. It's definitely not the right long-term direction, but we may end up not being able to support the same small storage on 32-bit as 64-bit.